### ISSUE TITLE
Backend strategy

### ISSUE THREAD

**Issue by mayou36** [zfit/zfit-developement#18]
@mayou36 commented on [Fri May 18 2018](https://github.com/zfit/zfit/issues/21)

We currently fully rely on TF. In the foreseeable future, this may change, so the backend should desirably be exchangeable. I came up with three possible ways to go, each with their own advantages and disadvantages.

## 1 Backend wrapping on low level (like Keras does)
we could in principle wrap tensorflow, pytorch on a low level and use this functions for operations (like `dot` instead of `tf.matmul` or `torch.mm` etc.).  

#### Advantage: 
- _once_ implemented, it yields full flexibility!

#### Disadvantage:
- a lot of work! Example: [keras wrapped tf backend](https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py)  
- yet another new API to learn for users already used to tf or torch

## 2 Strategy pattern
Simple: a pdf contains several implementations of the pdf (with tf, torch). Depending on the backend, a different implementation is chosen each time (of course, automatically).

#### Advantage:
- Allows easily even for mixed backends (in theory), allows flexible backend change
- Somehow beautiful solution :)
#### Disadvantage:
- each object has to be taken care of individually: this works nice for, say, pdfs, but get's fast out of hands when implementing `Variable` and does not work for static functions (or not so easily).

## 3 Name mangling on import time
Knowing the backend, on import time the right module is made available in the namespace. E.g. there is a `sum_tf`, `sum_torch` function and, depending on the backend chosen, either `sum = sum_tf` or `sum = sum_torch` leads to `from foo import sum` always yielding the right function.

#### Advantage:
- super easy, though rather brute force and inelegant implementation
#### Disadvantage:
- static: mixing not possible (or with backend knowledge required), flexible change not possible (first set backend, then do imports)

# Conclusion
1 is unrealistic. 2 can get out of hands with complexity. So realistically speaking, I think we will go for 3. The backend is not an important part and "could" even be just fixed. All we need to make sure is therefore to have

- [ ] a backend flag that can be set without other imports
- [ ] a function (or some kind of strategy) to mangle the names ("worst" case: simple ifs)

What are your thoughts about that @apuignav ?

---

@apuignav commented on [Fri May 18 2018](https://github.com/zfit/zfit/issues/21#issuecomment-390260863)

This is very nicely summarized. I thought 2 would be easy, but now that you say it you're right, it's quite easy that it gets out of hand. How would we go out to set up the backend before the imports? With some sort of `setup` function? It's a bit unclear to me how to do it in a reasonably elegant way.

---

@mayou36 commented on [Sat May 19 2018](https://github.com/zfit/zfit/issues/21#issuecomment-390425693)

(I would love 2, but for a whole repo: Parameter, phase-space...)

For 1: That's actually not too hard. Simple: config file (.zfit in home). And we may even provide it as `zfit.backend = 'tf'` or similar. Internally, this works as follows, for example by using lazy imports so that only the explicit attribute access `zfit.foo`  triggers the import of `foo` inside `zfit.__init__.py`. Or that changing zfit.backend triggers a recursive reload inside zfit.

And an example of the name mangling:
```
foo.py

__all__ == ['bar']

def bar_tf():
	body
def bar_torch():
	body

if backend == 'tf':
	bar = bar_tf
elif backend == 'torch':
	bar = bar_torch
```

or in the __init__.py at foo.py with `import .foo.bar_{tf,torch} as bar`

Both not very elegant, but working and not hard to implement (and sure, this can be made even more elegant, sticking to a fixed naming convention etc.)
